tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Serializers/KnownSerializers/KnownSerializerFinderTests.cs

```diff
@@ -0,0 +1,222 @@
+        }
+
+        [Fact]
+        public void Conditional_expression_with_different_enum_representations_should_throw()
+        {
+            Expression<Func<C, E>> expression = x => x.Ei == E.A ? E.B : x.Es;
+            var collectionSerializer = GetCollectionSerializer();
+
+            var result = KnownSerializerFinder.FindKnownSerializers(expression, collectionSerializer);
+
+            var conditionalExpression = (ConditionalExpression)expression.Body;
+            Assert.Throws<InvalidOperationException>(() => result.GetSerializer(conditionalExpression.IfTrue));
+        }
+
+        [Fact]
+        public void Property_chain_should_return_correct_nested_serializer()
+        {
+            Expression<Func<C, B>> expression = x => x.A.B;
+            var collectionSerializer = GetCollectionSerializer();
+
+            var result = KnownSerializerFinder.FindKnownSerializers(expression, collectionSerializer);
+
+            var serializer = result.GetSerializer(expression.Body);
+            collectionSerializer.TryGetMemberSerializationInfo(nameof(C.A), out var aSerializationInfo).Should().BeTrue();
+            ((BsonClassMapSerializer<A>)aSerializationInfo.Serializer).TryGetMemberSerializationInfo(nameof(A.B), out var bSerializationInfo).Should().BeTrue();
+            serializer.Should().Be(bSerializationInfo.Serializer);
+        }
+
+        [Fact]
+        public void Two_property_chains_should_each_return_correct_nested_serializer()
+        {
+            Expression<Func<C, E>> expression = x => x.A.B == null ? x.Ei : x.Es;
+            var collectionSerializer = GetCollectionSerializer();
+
+            var result = KnownSerializerFinder.FindKnownSerializers(expression, collectionSerializer);
+
+            var conditionalExpression = (ConditionalExpression)expression.Body;
+            var trueBranchSerializer = result.GetSerializer(conditionalExpression.IfTrue);
+            collectionSerializer.TryGetMemberSerializationInfo(nameof(C.Ei), out var eiSerializationInfo).Should().BeTrue();
+            trueBranchSerializer.Should().Be(eiSerializationInfo.Serializer);
+            var falseBranchSerializer = result.GetSerializer(conditionalExpression.IfFalse);
+            collectionSerializer.TryGetMemberSerializationInfo(nameof(C.Es), out var esSerializationInfo).Should().BeTrue();
+            falseBranchSerializer.Should().Be(esSerializationInfo.Serializer);
+        }
+
+        [Fact]
+        public void Projection_into_new_type_should_return_correct_serializer()
+        {
+            Expression<Func<C, View>> expression = x => new View { A = x.A };
+            var collectionSerializer = GetCollectionSerializer();
+
+            var result = KnownSerializerFinder.FindKnownSerializers(expression, collectionSerializer);
+
+            var serializer = result.GetSerializer(expression.Body);
+            serializer.ValueType.Should().Be(typeof(View));
```

Comment:
Doesn't seem to be asserting enough.

Maybe it's all we can do though?

Reply:
I've improved the asserts to verify that we get back the expected type of serializer and that the serializer for `View.A` is the same as for `C.A`.